File: python/examples/src/diffusion_examples/messaging/message_to_message_path.py
"""
Copyright © 2024 Diffusion Data Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import asyncio
import typing

from diffusion import sessions, Credentials, SessionId
from diffusion.messaging import RequestHandler
from diffusion_examples.utils.program import Example
import diffusion.datatypes


class MessageToMessagePath(Example):
    async def run(
        self,
        server_url: str = "<url>",
        principal: str = "<principal>",
        password: str = "<password>",
    ) -> None:
        path = "my/message/path"

        async with sessions().principal(principal).credentials(
            Credentials(password)
        ).open(server_url) as session:
            request_handler = SimpleRequestHandler()
            await session.messaging.add_request_handler(path, request_handler)

            async with sessions().principal(principal).credentials(
                Credentials(password)
            ).open(server_url) as session2:
                response = await session2.messaging.send_request_to_path(
                    path, diffusion.datatypes.STRING("Hello")
                )
                print(f"Received response: {response}.")


class SimpleRequestHandler(RequestHandler):
    def __init__(self) -> None:
        super().__init__(
            self.on_request,
            diffusion.datatypes.STRING,
            diffusion.datatypes.STRING,
        )

    # noinspection PyUnusedLocal
    async def on_request(
        self,
        request: typing.Optional[str],
        *,
        sender_session_id: typing.Optional[SessionId] = None,
        path: typing.Optional[str] = None,
        **kwargs,
    ) -> str:
        print(f"Received message: {request}.")
        return "Goodbye"



if __name__ == "__main__":
    asyncio.run(
        MessageToMessagePath().run(
            server_url="ws://localhost:8080",
            principal="admin",
            password="password",
        )
    )

File: dotnet/examples/Messaging/MessageToMessagePath.cs
/**
 * Copyright © 2023 - 2024 Diffusion Data Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Threading;
using System.Threading.Tasks;
using PushTechnology.ClientInterface.Client.Callbacks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Program;

namespace PushTechnology.ClientInterface.Examples.Messaging
{
    public sealed class MessageToMessagePath : Example {
        public override async Task Run( CancellationToken cancellationToken, string[] args ) 
        {
            string serverUrl = args[0];

            var session = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            string path = "my/message/path";

            var requestHandler = new SimpleRequestHandler();
            var registration = await session.Messaging.AddRequestHandlerAsync<string, string>(
                path, requestHandler, cancellationToken);

            var session2 = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .Open(serverUrl);

            string response = await session2.Messaging.SendRequestAsync<string, string>(
                path, "Hello", cancellationToken);
            WriteLine($"Received response: {response}.");

            session.Close();
            session2.Close();
        }

        private class SimpleRequestHandler : IRequestHandler<string, string> {
            public void OnClose() {}

            public void OnError(ErrorReason errorReason) {}

            public void OnRequest( string request, IRequestContext context, IResponder<string> responder ) {
                WriteLine( $"Received message: {request}." );

                responder.Respond( "Goodbye" );
            }
        }
    }
}
